app/models/cas/non_hmis_client.rb

```diff
@@ -0,0 +1,124 @@
+    self.inheritance_column = :_type_disabled # disable STI
+
+    belongs_to :client, class_name: 'GrdaWarehouse::Hud::Client', foreign_key: :warehouse_client_id, optional: true
+
+    # Although default_scope is generally discouraged, one of the reasons CAS is a separate DB, is to keep
+    # Deidentified clients out of the warehouse...
+    default_scope { where(identified: true) }
+
+    scope :unassigned, -> do
+      where(warehouse_client_id: nil)
+    end
+
+    MATCH_COLUMNS = [:first_name, :last_name, :date_of_birth, :ssn].freeze
+
+    def self.find_matches(batch)
+      candidates = batch.pluck(:id, *MATCH_COLUMNS).
+        map { |id, *rest| [id, MATCH_COLUMNS.zip(rest).to_h] }.
+        to_h
+
+      candidates.each do |_, candidate|
+        candidate[:first_name] = candidate[:first_name].downcase.strip.gsub(/[^a-z0-9]/i, '')
+        candidate[:last_name] = candidate[:last_name].downcase.strip.gsub(/[^a-z0-9]/i, '')
+        candidate[:ssn] = candidate[:ssn].strip.gsub(/[^0-9]/, '')
+      end
+
+      by_first_name = candidates.reject { |_, h| h[:first_name].blank? }.group_by { |_, h| h[:first_name] }.transform_values { |c| c.flatten.first }
+      by_last_name = candidates.reject { |_, h| h[:last_name].blank? }.group_by { |_, h| h[:last_name] }.transform_values { |c| c.flatten.first }
+      by_dob = candidates.reject { |_, h| h[:date_of_birth].blank? }.group_by { |_, h| h[:date_of_birth] }.transform_values { |c| c.flatten.first }
+      by_ssn = candidates.reject { |_, h| h[:ssn].blank? }.group_by { |_, h| h[:ssn] }.transform_values { |c| c.flatten.first }
+
+      first_name_matches = client_source.where(c_t[:FirstName].lower.in(by_first_name.keys)).
+        pluck(c_t[:FirstName].lower, :id).
+        group_by(&:first).
+        transform_values { |v| v.map(&:last) }
+      last_name_matches = client_source.where(c_t[:LastName].lower.in(by_last_name.keys)).
+        pluck(c_t[:LastName].lower, :id).
+        group_by(&:first).
+        transform_values { |v| v.map(&:last) }
+      dob_matches = client_source.where(DOB: by_dob.keys).
+        pluck(:DOB, :id).
+        group_by(&:first).
+        transform_values { |v| v.map(&:last) }
+      ssn_matches = client_source.where(SSN: by_ssn.keys).
+        pluck(:SSN, :id).
+        group_by(&:first).
+        transform_values { |v| v.map(&:last) }
+
+      OpenStruct.new(
+        candidates: candidates,
+        first_name_matches: first_name_matches,
+        last_name_matches: last_name_matches,
+        dob_matches: dob_matches,
+        ssn_matches: ssn_matches,
+      )
+    end
+
+    # An "exact match" is actually 3 out of 4 items
+    def self.exact_match(candidate, matches)
+      name_matches = (matches.first_name_matches[candidate[:first_name]] || []) & (matches.last_name_matches[candidate[:last_name]] || [])
+      exact_matches = name_matches & matches.dob_matches[candidate[:date_of_birth]]
+      exact_matches += name_matches & matches.ssn_matches[candidate[:ssn]]
```

Comment:
This doesn't feel quite right to me.  There's no obvious check to confirm that the two of the three match.  I think you want to get the ids where the name matches, the ids where the dob matches, the ids where the ssn matches, and then reject any that have fewer than 2 instances in the array.
As written, it looks like name has to match, then either dob or ssn has to match.  But having a matching DOB and SSN would not be sufficient.

Reply:
Also, it may be really hard to do, but I think we need a test or two for this as it will mess with the connections between CAS clients and Warehouse clients without human intervention.